src/entry.c

```diff
@@ -0,0 +1,385 @@
+    bool embed_value = false;
+    int embedded_field_sds_type;
+    size_t expiry_size, embedded_value_sds_size, embedded_field_sds_size;
+    size_t alloc_size = entryReqSize(field, value, expiry, &embed_value, &embedded_field_sds_type, &embedded_field_sds_size, &expiry_size, &embedded_value_sds_size);
+    size_t buf_size;
+
+    /* allocate the buffer */
+    char *buf = zmalloc_usable(alloc_size, &buf_size);
+
+    /* Set The expiry if exists */
+    if (expiry_size) {
+        *(long long *)buf = expiry;
+        buf += expiry_size;
+        buf_size -= expiry_size;
+    }
+    if (value) {
+        if (!embed_value) {
+            *(sds *)buf = value;
+            buf += sizeof(sds);
+            buf_size -= sizeof(sds);
+        } else {
+            sdswrite(buf + embedded_field_sds_size, buf_size - embedded_field_sds_size, SDS_TYPE_8, value, sdslen(value));
+            sdsfree(value);
+            buf_size -= embedded_value_sds_size;
+        }
+    }
+    /* Set the field data */
+    entry *new_entry = sdswrite(buf, embedded_field_sds_size, embedded_field_sds_type, field, sdslen(field));
+
+    /* Field sds aux bits are zero, which we use for this entry encoding. */
+    sdsSetAuxBit(new_entry, FIELD_SDS_AUX_BIT_ENTRY_HAS_VALUE_PTR, embed_value ? 0 : 1);
+    sdsSetAuxBit(new_entry, FIELD_SDS_AUX_BIT_ENTRY_HAS_EXPIRY, expiry_size > 0 ? 1 : 0);
+
+    /* Check that the new entry was built correctly */
+    debugServerAssert(sdsGetAuxBit(new_entry, FIELD_SDS_AUX_BIT_ENTRY_HAS_VALUE_PTR) == (embed_value ? 0 : 1));
+    debugServerAssert(sdsGetAuxBit(new_entry, FIELD_SDS_AUX_BIT_ENTRY_HAS_EXPIRY) == (expiry_size > 0 ? 1 : 0));
+    return new_entry;
+}
+
+/* Modify the entry's value and/or expiration time.
+ * In case the provided value is NULL, will use the existing value. */
+entry *entryUpdate(entry *e, sds value, long long expiry) {
+    sds field = (sds)e;
+    entry *new_entry = NULL;
+
+    bool update_value = value ? true : false;
+    long long curr_expiration_time = entryGetExpiry(e);
+    bool update_expiry = (expiry != curr_expiration_time) ? true : false;
+    /* Just a sanity check. If nothing changes, lets just return */
+    if (!update_value && !update_expiry)
+        return e;
+
+    if (!value) value = entryGetValue(e);
+    bool embed_value = false;
+    int embedded_field_sds_type;
+    size_t expiry_size, embedded_value_size, embedded_field_size;
+    size_t required_embedded_size = entryReqSize(field, value, expiry, &embed_value, &embedded_field_sds_type, &embedded_field_size, &expiry_size, &embedded_value_size);
+    size_t current_embedded_allocation_size = entryHasValuePtr(e) ? 0 : entryMemUsage(e);
+
+    bool expiry_add_remove = update_expiry && (curr_expiration_time == EXPIRY_NONE || expiry == EXPIRY_NONE); // In case we are toggling expiration
+    bool value_change_encoding = update_value && (embed_value != !entryHasValuePtr(e));                       // In case we change the way value is embedded or not
```

Comment:
double negation is hard to read
```suggestion
    bool value_change_encoding = update_value && (embed_value != entryHasEmbeddedValue(e));                       // In case we change the way value is embedded or not
```

Reply:
hhh. I only added the entryHasEmbeddedValue for tests, but you are right.